lib/blockchain.go

```diff
@@ -487,14 +491,52 @@ type Blockchain struct {
 	timer *Timer
 }
 
-func (bc *Blockchain) CopyBlockIndex() map[BlockHash]*BlockNode {
-	newBlockIndex := make(map[BlockHash]*BlockNode)
-	for kk, vv := range bc.blockIndex {
-		newBlockIndex[kk] = vv
+func (bc *Blockchain) addNewBlockNodeToBlockIndex(blockNode *BlockNode) {
+	bc.blockIndexByHash[*blockNode.Hash] = blockNode
+	if _, exists := bc.blockIndexByHeight[uint64(blockNode.Height)]; !exists {
+		bc.blockIndexByHeight[uint64(blockNode.Height)] = make(map[BlockHash]*BlockNode)
+	}
+	bc.blockIndexByHeight[uint64(blockNode.Height)][*blockNode.Hash] = blockNode
+}
+
+func (bc *Blockchain) CopyBlockIndexes() (_blockIndexByHash map[BlockHash]*BlockNode, _blockIndexByHeight map[uint64]map[BlockHash]*BlockNode) {
+	newBlockIndexByHash := make(map[BlockHash]*BlockNode)
+	newBlockIndexByHeight := make(map[uint64]map[BlockHash]*BlockNode)
+	for kk, vv := range bc.blockIndexByHash {
+		newBlockIndexByHash[kk] = vv
+		blockHeight := uint64(vv.Height)
+		if _, exists := newBlockIndexByHeight[blockHeight]; !exists {
+			newBlockIndexByHeight[blockHeight] = make(map[BlockHash]*BlockNode)
+		}
+		newBlockIndexByHeight[blockHeight][kk] = vv
+	}
+	return newBlockIndexByHash, newBlockIndexByHeight
+}
+
+func (bc *Blockchain) ConstructBlockIndexByHeight() map[uint64]map[BlockHash]*BlockNode {
+	newBlockIndex := make(map[uint64]map[BlockHash]*BlockNode)
+	for _, blockNode := range bc.blockIndexByHash {
+		blockHeight := uint64(blockNode.Height)
+		if _, exists := newBlockIndex[blockHeight]; !exists {
+			newBlockIndex[blockHeight] = make(map[BlockHash]*BlockNode)
+		}
+		newBlockIndex[blockHeight][*blockNode.Hash] = blockNode
 	}
 	return newBlockIndex
 }
 
+func (bc *Blockchain) getAllBlockNodesByHeight(blockHeight uint64) []*BlockNode {
```

Comment:
Just spitballing here. WDYT of this name so it's clear we're checking the block index? This would be consistent to the naming for the helper functions above too
```suggestion
func (bc *Blockchain) getAllBlocksIndexedAtHeight(blockHeight uint64) []*BlockNode {
```

Reply:
I went w/ `getAllBlockNodesIndexedAtHeight` so it's clear this function returns BlockNode structs.